src/pubd/repository.rs

```diff
@@ -556,25 +556,41 @@ impl RrdpServer {
         }
     }
 
-    // Push the delta and truncate excessive deltas.
-    //  - keep no more than the size of the snapshot
-    //  - don't keep deltas older than configured seconds
-    //  - but keep a minimum of configured deltas
+    // Push the delta and truncate excessive deltas:
+    //  - never keep more than the size of the snapshot
+    //  - always keep 'retention_delta_files_min_nr' files
+    //  - always keep 'retention_delta_files_min_seconds' files
+    //  - beyond this:
+    //     - never keep more than 'retention_delta_files_max_nr'
+    //     - never keep older than 'retention_delta_files_max_seconds'
+    //     - keep the others
     fn update_deltas(&mut self, elements: DeltaElements, config: &RepositoryRetentionConfig) {
         self.deltas.push_front(Delta::new(self.session, self.serial, elements));
         let mut keep = 0;
         let mut size = 0;
         let snapshot_size = self.snapshot.size();
 
-        let retain_secs = config.retention_delta_files_seconds;
-        let minimum = config.retention_delta_files_min_nr;
+        let min_nr = config.retention_delta_files_min_nr;
+        let min_secs = config.retention_delta_files_min_seconds;
+        let max_nr = config.retention_delta_files_max_nr;
+        let max_secs = config.retention_delta_files_max_seconds;
 
         for delta in &self.deltas {
             size += delta.elements().size();
-            if size > snapshot_size || (keep > minimum && delta.older_than_seconds(retain_secs)) {
-                break;
+
+            if size > snapshot_size {
+                // never keep more than the size of the snapshot
+            } else if keep < min_nr || delta.younger_than_seconds(min_secs) {
+                // always keep 'retention_delta_files_min_nr' files
+                // always keep 'retention_delta_files_min_seconds' file
+                keep += 1
+            } else if keep == max_nr || delta.older_than_seconds(max_secs) {
```

Comment:
Is self.deltas ordered by age? If so you break out of the loop here. If not, what happens if one delta is older sthan max seconds so is skipped but we go round the loop again and find the next delta is _younger than max seconds and so is kept, but now keep has incremented by 1 referring to keeping the last delta, not this one??? This looks correct but feels like there is potential for errors here.

Reply:
yep, now breaking